source/d3d11/d3d11_device.cpp

```diff
@@ -510,3 +513,47 @@ void STDMETHODCALLTYPE D3D11Device::ReadFromSubresource(void *pDstData, UINT Dst
 
 	static_cast<ID3D11Device3 *>(_orig)->ReadFromSubresource(pDstData, DstRowPitch, DstDepthPitch, pSrcResource, SrcSubresource, pSrcBox);
 }
+
+// performs a clean up of all the counter data
+void D3D11Device::perform_counterdata_cleanup()
+{
+	const std::lock_guard<std::mutex> lock(_counters_per_commandlist_mutex);
+	for (auto commandcounters : _counters_per_commandlist)
+	{
+		commandcounters.first->Release();
+		commandcounters.second.reset();
+	}
+	_counters_per_commandlist.clear();
+}
+
+// merges the counters logged for the specified command list in the counters destination tracker specified
+void D3D11Device::merge_commandlist_counters_in_counter_map(ID3D11CommandList* command_list, depth_counter_tracker& counters_destination)
+{
+	const std::lock_guard<std::mutex> lock(_counters_per_commandlist_mutex);
+
+	auto counters_for_commandlist = _counters_per_commandlist.find(command_list);
+	if (counters_for_commandlist == _counters_per_commandlist.end())
+	{
+		return;
+	}
+	counters_destination.merge(counters_for_commandlist->second);
+}
+
+// Merges the counters in counters_source in the counters_per_commandlist for the commandlist specified.
+void D3D11Device::merge_counters_per_commandlist(ID3D11CommandList* command_list, depth_counter_tracker& counters_source)
+{
+	const std::lock_guard<std::mutex> lock(_counters_per_commandlist_mutex);
+
+	auto entry = _counters_per_commandlist.find(command_list);
+	if (entry == _counters_per_commandlist.end())
+	{
+		depth_counter_tracker counters_per_depthstencil;
+		counters_per_depthstencil.merge(counters_source);
+		command_list->AddRef();
```

Comment:
I wouldn't increase the reference count here. It's not an issue if the pointer goes dangling because the game deleted the command list, since it's not used anywhere and is just there for lookup purposes. That way we don't have to deal with cleaning up the table either.

Reply:
I'll remove the addref. I added it in my quest to get rid of the memory leaks I discovered. I indeed found it unnecessary to use refcounting here (and also with the depthstencil as these are also just for lookup) but didn't know what unordered_map would do with the pointer under the hood, so to be safe I added the ref counter. 
Cleanup is still needed, as each frame commandlists are created and the table otherwise will increase each frame and lists which will never be used will stay in it.